### ISSUE TITLE
Support Java 8 Path API to allow using non-default filesystems

### ISSUE THREAD

**Issue by bannmann** [eclipse-jgit/jgit#58]
### Description

JGit should support the Java 8 `Path` API as follows:

- All methods that accept `java.io.File` instances would be overloaded to also accept `java.nio.file.Path`.
- Methods that take `File` arguments would be kept, but would use `path.toFile()` and call their `Path` counterparts.
- Internally, JGit would only use the `Path` API.

### Motivation

While writing a server application that clones repos, performs commits and pushes them, I encountered problems when trying to delete the repos. It seems files are locked for a (short?) time after JGit performed an operation.

If JGit supported the Path API, one could solve this problem by instantiating an in-memory Java filesystem like [Jimfs](https://github.com/google/jimfs) that only exists for the duration of the JGit operations (using a try-with-resources block):
```java
try (var fileSystem = Jimfs.newFileSystem(Configuration.unix()))
{
    Path repoDirectory = fileSystem.getPath("/repo");
    var git = cloneRepository(repoDirectory);
    // perform operations like commit or push
}
```

Other situations where this would offer advantages include:
- When disk I/O is a bottleneck: with Jimfs, there is no I/O.
- When mechanisms against filling up the hard disk / docker volume add too much complexity: with Jimfs, one thread potentially gets `OutOfMemoryErrors`, but the application keeps running. But with a real disks, all it takes is one too-large operation (or too many failed cleanups) - and the entire system comes to a halt!

### Alternatives considered

- Set up a cronjob outside Java (e.g. "once per day, delete all directories older than a day")
- After completing a repo operation, schedule a cleanup task (e.g. "in 6 hours, delete this repo directory")

Both workarounds bring their own caveats and complexities.

**Comment by msohn** [comment]
How did you create the `Git` instance ? Note that if creating it using the `Git(Repository)` constructor or the static `Git.wrap(Repository)` method `Git.close()` will not close the wrapped `Repository` as mentioned in the javadoc.

If you are creating the `Git` instance using one of the static `Git.open` methods the Git object uses the RepositoryCache and by default evicts the repository from WindowCache and RepositoryCache using a time based eviction strategy to prevent unnecessary cache thrashing of WindowCache and RepositoryCache which may degrade performance in long running server applications. By default eviction is handled by a background thread. 

There are 2 jgit options which can be used to configure this behavior:
- core.repositoryCacheExpireAfter: cache entries are evicted if the
cache entry wasn't accessed longer than this time in milliseconds
- core.repositoryCacheCleanupDelay: defines the interval in milliseconds
for running a background task evicting expired cache entries. If set to
-1 the delay is set to min(repositoryCacheExpireAfter, 10 minutes). If
set to 0 the time based eviction is switched off and no background task
is started. If time based eviction is switched off the JVM can still
evict cache entries if heap memory is running low.

See https://eclipse.gerrithub.io/c/eclipse-jgit/jgit/+/74904

If

**Comment by msohn** [comment]
Debug closing of the `Git` instance and check in `Repository.close()` if the ref count dropped to 0.

**Comment by bannmann** [comment]
@msohn
Yes, inside `Repository.close()`, the `newCount` is `0`. As `RepositoryCache.isCached(this)` is `false`, it then calls `doClose()`.

The `java.nio.file.AccessDeniedException` I get is about `<repodir>\.git\objects\84\0eebb15d72c5cf67da746f218d471580b49b15` (I'm on a Windows machine). The concrete file name is different for each run (presumably because it relates to the commit that was just made, which obviously has a different hash).

When I remove the git operations so that after cloning so that the `try` block literally does nothing with the `Git` instance or the directory, something interesting happens: deleting the directory _still_ results in an `AccessDeniedException`. However, different to the case above, it's about an idx file: `<repodir>\.git\objects\pack\pack-fe4c4220aac672204f1b45bd9c01d5bb8e3516b5.idx`. Also, that file name is always the same (even when new commits have been added in the meantime and therefore appear when cloning the repo in the next run). I don't know whether any of that is relevant.

By the way, this happens with JGit 6.9.0.202403050737-r.

**Comment by msohn** [comment]
Do you keep any reference to some jgit object which has a close method you didn't call before you try to delete the repo ?

You can print the content of the loose object using `git cat-file -p <path of the loose object>`. Maybe this gives you a clue.

`Repository.doClose()` calls `ObjectDirectory.close` which should close all packs and clears the loose object cache.
Pack indexes are read in `PackIndex.open()` on first access using a file inputstream constructed in a try with resource (so it's guaranteed to be closed), the PackIndex object is cached in memory until it's cleared when `Pack.close()` is called.
New objects added to the repo by creating new commits don't immediately end up in pack files, they are initially stored as loose objects and are later eventually packed by a run of `git gc`.

**Comment by sschuberth** [comment]
> But back to the original topic of supporting the Path API: I'd also be interested in using JGit with Jimfs to quickly create in-memory Git repos for testing Git-specific code in tests.

I just realized that JGit itself provides an [InMemoryRepository](https://github.com/eclipse-jgit/jgit/blob/48da42bbf7aae7555ccf813af1b04fc2625befbe/org.eclipse.jgit/src/org/eclipse/jgit/internal/storage/dfs/InMemoryRepository.java#L40), which probably renders the need for Jimfs void.
